### PR TITLE
Add webhooks subscription try out to devportal.

### PR DESCRIPTION
**Pull request by sachithKay**
- Removes existing webhook subscription URL view.
- Adds a new try out view for webhooks API's

Screen Shot 2021-03-05 at 3 51 43 PM

Screen Shot 2021-03-05 at 3 52 09 PM

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+                    const scopeList = apiData.scopes.map((scope) => {
+                        return scope.name;
+                    });
+                    setScopes(scopeList);
+                }
+            })
+            .catch((error) => {
+                if (process.env.NODE_ENV !== 'production') {
+                    console.error(error);
+                }
```
Comment (chanaka3d) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx:109]: It's better to show an error to the user here as well.
```
Alert.error(error);
```

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+    if (api && api.securityScheme) {
+        isApiKeyEnabled = api.securityScheme.includes('api_key');
+        if (isApiKeyEnabled && securitySchemeType === 'API-KEY') {
+            authorizationHeader = 'apikey';
+        }
+    }
+
+    const isPrototypedAPI = api.lifeCycleStatus && api.lifeCycleStatus.toLowerCase() === 'prototyped';
+    return (
```
Comment (chanaka3d) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx:183]: Shall we set the initial state of scopeList, Label, environment, to null and show loading while the response comes. 
```javascript
if(!scopeList || label || environment) {
return ( <Progress /> );
}
```

**Review by chanaka3d — APPROVED** [review]

**Review by tharindu1st — APPROVED** [review]

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx (hunk 1 above)
Comment (sachithKay) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx:111]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx (hunk 2 above)
Comment (sachithKay) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx:188]: else part was handled in the catch as discussed

**Review by chanaka3d — APPROVED** [review]

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                return sandboxApiKey;
+            }
+        } else if (selectedKeyType === 'PRODUCTION') {
+            return productionAccessToken;
+        } else {
+            return sandboxAccessToken;
+        }
+    }
+
+    if (api == null || !scopes || !labels || !environments) {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx:169]: this should be api `===` null

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        return <Progress />;
+    }
+    if (notFound) {
+        return 'API Not found !';
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx:173]: Should internationalize the text

Comment (tmkasun) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx:173]: Using `FormattedMessage` component

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+            padding: 0,
+            'label + &': {
+                marginTop: theme.spacing(1),
+            },
+        },
+        bootstrapInput: {
+            borderRadius: 4,
+            backgroundColor: theme.palette.common.white,
+            border: '1px solid #ced4da',
+            padding: '5px 12px',
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:49]: It'ss good to use `Box` component with default padding margin units

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            '&:focus': {
+                borderColor: '#80bdff',
+                boxShadow: '0 0 0 0.2rem rgba(0,123,255,.25)',
+            },
+            fontSize: 12,
+        },
+        bootstrapCurl: {
+            borderRadius: 4,
+            backgroundColor: '#41444e',
+            color: '#fff',
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:63]: Color values should come from the theme file

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    const classes = useStyles();
+    const { generateGenericWHSubscriptionCurl, topic, intl } = props;
+    const initialSubscriptionState = {
+        topic: topic.name,
+        secret: null,
+        lease: 50000,
+        mode: 'subscribe',
+        callback: null,
+    };
+    const [curl, setCurl] = useState(generateGenericWHSubscriptionCurl(initialSubscriptionState));
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:106]: This kind of complex state favor to use `useReducer`

> useReducer is usually preferable to useState when you have complex state logic that involves multiple sub-values . . .

~[React Docs](https://reactjs.org/docs/hooks-reference.html#usereducer)

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    const [formError, setFormError] = useState(false);
+    const [state, dispatch] = useReducer(reducer, initialSubscriptionState);
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:108]: Why we need `useState` & `useReducer` both ?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx (hunk 7 above)
Comment (tmkasun) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:106]: ahh ok this state only holds the curl generated from the `initialSubscriptionState` 🤓

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx (hunk 8 above)
Comment (tmkasun) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:108]: Ok got it 👍

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+            setFormError(false);
+            setCurl(generateGenericWHSubscriptionCurl(state));
+        }
+    };
+
+    const handleChange = (e) => {
+        dispatch({ field: e.target.name, value: e.target.value });
+    };
+
+    function onCopy(message) {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:123]: Do we really need a function for this why can't we just call `Alert.info` right away :)

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        Alert.info(message);
+    }
+
+    return (
+        <Accordion className={classes.subscription}>
+            <AccordionSummary
+                expandIcon={<ExpandMoreIcon />}
+                aria-controls='panel1bh-content'
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:131]: Give a proper name to `aria-controls`

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+                        <TextField
+                            label={intl.formatMessage({
+                                defaultMessage: 'cURL',
+                                id: 'Apis.Details.AsyncApiConsole.Webhooks.curl',
+                            })}
+                            defaultValue=''
+                            value={curl}
+                            fullWidth
+                            multiline
+                            id='bootstrap-input'
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:244]: Do we need this id ?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+                            InputLabelProps={{
+                                shrink: true,
+                                className: classes.bootstrapFormLabel,
+                            }}
+                        />
+                    </Grid>
+                </Grid>
+            </AccordionDetails>
+            <AccordionActions style={{ paddingRight: '18px' }}>
+                <Button size='small' onClick={handleClick}>Generate Curl</Button>
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/WebhookSubscriptionUI.jsx:261]: Should support i18n `FormattedMessage`

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/index.jsx [hunk 13]
```diff
@@ -62,7 +63,9 @@ const LoadableSwitch = withRouter((props) => {
     let tryoutRoute;
     if (api.type === 'GRAPHQL') {
         tryoutRoute = <Route path='/apis/:apiUuid/test' component={GraphQLConsole} />
-    }else {
+    } else if (api.type === 'WS' || api.type === 'WEBSUB' || api.type === 'SSE') {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/index.jsx:66]: Shall we move these to constants ?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+import Alert from 'AppComponents/Shared/Alert';
+import TryOutController from '../ApiConsole/TryOutController';
+import { ApiContext } from '../ApiContext';
+import Api from '../../../../data/api';
+import Progress from '../../../Shared/Progress';
+import AsyncApiUI from './AsyncApiUI';
+
+const useStyles = makeStyles((theme) => ({
+    buttonIcon: {
+        marginRight: 10,
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Apis/Details/AsyncApiConsole/AsyncApiConsole.jsx:37]: Better to use

```jsx
<Box mr={1}>.......</Box>

```

instead

**Review by tmkasun — APPROVED** [review]

**Comment by tmkasun** [comment]
Please address the review comments and merge

**Comment by sachithKay** [comment]
merging upon addressing the review comments.

**Pull request merged**